streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBTimeOrderedKeyValueBuffer.java

```diff
@@ -0,0 +1,171 @@
+    }
+
+    @Deprecated
+    @Override
+    public void init(final ProcessorContext context, final StateStore root) {
+        wrapped().init(context, wrapped());
+    }
+
+    @Override
+    public void init(final StateStoreContext context, final StateStore root) {
+        wrapped().init(context, wrapped());
+    }
+
+    @Override
+    public void evictWhile(final Supplier<Boolean> predicate, final Consumer<Eviction<K, V>> callback) {
+        KeyValue<Bytes, byte[]> keyValue;
+
+        if (predicate.get()) {
+            try (final KeyValueIterator<Bytes, byte[]> iterator = wrapped()
+                .fetchAll(0, wrapped().observedStreamTime - gracePeriod)) {
+                while (iterator.hasNext() && predicate.get()) {
+                    keyValue = iterator.next();
+
+                    final BufferValue bufferValue = BufferValue.deserialize(ByteBuffer.wrap(keyValue.value));
+                    final K key = keySerde.deserializer().deserialize(topic,
+                        PrefixedWindowKeySchemas.TimeFirstWindowKeySchema.extractStoreKeyBytes(keyValue.key.get()));
+                    minTimestamp = bufferValue.context().timestamp();
+
+                    if (wrapped().observedStreamTime - gracePeriod > minTimestamp) {
+                        return;
+                    }
+
+                    V value = valueSerde.deserializer().deserialize(topic, bufferValue.newValue());
+
+                    if (bufferValue.context().timestamp() != minTimestamp) {
+                        throw new IllegalStateException(
+                            "minTimestamp [" + minTimestamp + "] did not match the actual min timestamp [" +
+                                bufferValue.context().timestamp() + "]"
+                        );
+                    }
+
+                    callback.accept(new Eviction<K, V>(key, value, bufferValue.context()));
+
+                    wrapped().remove(keyValue.key);
+                    numRecords--;
+                    bufferSize = bufferSize - computeRecordSize(keyValue.key, bufferValue);
+                }
+                if (numRecords == 0) {
+                    minTimestamp = Long.MAX_VALUE;
+                }
+            }
+        }
+    }
+
+
+    @Override
+    public Maybe<ValueAndTimestamp<V>> priorValueForBuffered(final K key) {
+        return Maybe.undefined();
+    }
+
+    @Override
+    public void put(final long time, final Record<K, V> record, final ProcessorRecordContext recordContext) {
+        requireNonNull(record.value(), "value cannot be null");
+        requireNonNull(recordContext, "recordContext cannot be null");
+        if (wrapped().observedStreamTime - gracePeriod > record.timestamp()) {
+            return;
+        }
+        final Bytes serializedKey = Bytes.wrap(
+            PrefixedWindowKeySchemas.TimeFirstWindowKeySchema.toStoreKeyBinary(keySerde.serializer().serialize(topic, record.key()),
+                record.timestamp(),
+                seqnum++).get());
```

Comment:
`RocksDBTimeOrderedWindowStore` does something similar. It starts at zero and updates the sequnum when inserting

Reply:
@cadonna is this also an issue or is there something else I am missing?

Reply:
+1 to Bruno's confusion here. I'm not sure how this works for `RocksDBTimeOrderedWindowStore` either.

Also, if we go with the approach of incrementing seqnum, we need to guard against overflow. (`RocksDBTimeOrderedWindowStore` does this in `maybeUpdateSeqnumForDups()`.)

Reply:
I'll add int the protection for overflow, but I will figure out the seqnum being reset when I add the recovery logic in a later PR. That will have tests for things like this anyways and I think this PR is already getting too big.

Reply:
We talked about this in an offline sync. This is a problem throughout all of Streams since its inception. We won't be fixing it here but we can follow up later with a general fix when it gets prioritized.